src/mqt/predictor/hellinger/utils.py

```diff
@@ -0,0 +1,134 @@
+    feature_dict.update(active_qubits_dict)
+
+    # Add the depth of the quantum circuit to the feature dictionary
+    feature_dict["depth"] = qc.depth()
+
+    # Add the number of active qubits to the feature dictionary
+    num_active_qubits = sum(active_qubits_dict.values())
+    feature_dict["num_qubits"] = num_active_qubits  # NOTE: not used in feature vector
+
+    # Calculate supermarq features, which uses circ.num_qubits
+    assert device.num_qubits == qc.num_qubits
+    supermarq_features = calc_supermarq_features(qc)
+    feature_dict["critical_depth"] = supermarq_features.critical_depth
+    feature_dict["entanglement_ratio"] = supermarq_features.entanglement_ratio
+    feature_dict["parallelism"] = supermarq_features.parallelism
+    feature_dict["liveness"] = supermarq_features.liveness
+
+    # NOTE: renormalize using the number of active qubits (not the device's num_qubits)
+    feature_dict["parallelism"] = (
+        feature_dict["parallelism"] * (device.num_qubits - 1) / (num_active_qubits - 1) if num_active_qubits >= 2 else 0
+    )
+    feature_dict["liveness"] = (
+        feature_dict["liveness"] * device.num_qubits / num_active_qubits if num_active_qubits >= 1 else 0
+    )
+
+    # Calculate additional features based on DAG
+    dag = circuit_to_dag(qc)
+    dag.remove_all_ops_named("barrier")
+    dag.remove_all_ops_named("measure")
+
+    # Directed program communication = circuit's average directed qubit degree / degree of a complete directed graph.
+    di_graph = nx.DiGraph()
+    for op in dag.two_qubit_ops():
+        q1, q2 = op.qargs
+        di_graph.add_edge(qc.find_bit(q1).index, qc.find_bit(q2).index)
+    degree_sum = sum(di_graph.degree(n) for n in di_graph.nodes)
+    directed_program_communication = (
+        degree_sum / (2 * num_active_qubits * (num_active_qubits - 1)) if num_active_qubits >= 2 else 0
+    )
+
+    # Average number of 1q gates per layer = num of 1-qubit gates in the circuit / depth
+    single_qubit_gates_per_layer = (
+        (len(dag.gate_nodes()) - len(dag.two_qubit_ops())) / dag.depth() if dag.depth() > 0 else 0
+    )
+    # Average number of 2q gates per layer = num of 2-qubit gates in the circuit / depth
+    multi_qubit_gates_per_layer = len(dag.two_qubit_ops()) / dag.depth() if dag.depth() > 0 else 0
+
+    # Normalize both values using the available number of qubits
+    single_qubit_gates_per_layer = single_qubit_gates_per_layer / num_active_qubits if num_active_qubits > 1 else 0
+    multi_qubit_gates_per_layer = multi_qubit_gates_per_layer / (num_active_qubits // 2) if num_active_qubits > 2 else 0
+
+    feature_dict["directed_program_communication"] = directed_program_communication
+    feature_dict["single_qubit_gates_per_layer"] = single_qubit_gates_per_layer
+    feature_dict["multi_qubit_gates_per_layer"] = multi_qubit_gates_per_layer
+
+    return np.array(list(feature_dict.values()))
+
+
+def get_hellinger_model_path(device: Device = None) -> Path:
```

Comment:
I am surprised that `mypy` is not complaining here: device defaults to `None` although it can only be of type `Device`.